File: samples/go/authenticate.go
package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha1"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
        "net/url"
	"os"
)

type keydata struct {
	SharedKey string
	ApiKey    string
}

func main() {
	signedURL := signURL("https://api.silobreaker.com/search/documents?q=Sweden&type=json")
	resp, e := http.Get(signedURL)
	if e != nil {
		fmt.Printf("Response error: %v\n", e)
		os.Exit(1)
	}
	defer resp.Body.Close()
	body, e := ioutil.ReadAll(resp.Body)
	if e != nil {
		fmt.Printf("Body error: %v\n", e)
		os.Exit(1)
	}

	fmt.Print(prettyJSON(body))
}

func prettyJSON(body []byte) string {

	var prettyJSON bytes.Buffer
	e := json.Indent(&prettyJSON, body, "", "  ")
	if e != nil {
		fmt.Printf("JSON error: %v\n", e)
		os.Exit(1)
	}

	return string(prettyJSON.Bytes())
}

func getKeys(keyfile string) keydata {
	file, e := ioutil.ReadFile(keyfile)
	if e != nil {
		fmt.Printf("File error: %v\n", e)
		os.Exit(1)
	}
	var keys keydata
	json.Unmarshal(file, &keys)
	return keys
}

func signURL(message string) string {
	keys := getKeys("../secrets.json")
	completeMessage := "GET " + message
	mac := hmac.New(sha1.New, []byte(keys.SharedKey))
	mac.Write([]byte(completeMessage))
	expectedMAC := mac.Sum(nil)
	digest := base64.StdEncoding.EncodeToString(expectedMAC)
	finalURL := message + "&apiKey=" + keys.ApiKey + "&digest=" + url.QueryEscape(digest)
	return finalURL
}

File: samples/python/authenticate.py
import base64
import hashlib
import hmac
import html
import json
import urllib.request
import argparse

# Command line arguments 

parser = argparse.ArgumentParser()
parser.add_argument("URL", help="the endpoint of the API, inside quotation marks")
parser.add_argument("-P", "--POST", help="perform a POST request. Data can be modified in post_data.json", action='store_true')
args = parser.parse_args()

url = args.URL

with open("../secrets.json") as f: # The secrets file has the same format as the node example.
    secrets = json.load(f)

sharedKey = secrets["SharedKey"]
apiKey = secrets["ApiKey"]

if args.POST:
    verb = "POST"
    with open('post_data.json', 'rb') as f:
        body = f.read()

    # Sign the URL
    urlSignature = verb + " " + url
    message = urlSignature.encode() + body

    hmac_sha1 = hmac.new(sharedKey.encode(), message, digestmod=hashlib.sha1)
    digest = base64.b64encode(hmac_sha1.digest())

    # Fetch the data

    final_url = url + "?apiKey=" + apiKey + "&digest=" + urllib.parse.quote(digest.decode())
    req = urllib.request.Request(final_url, data=body, headers={'Content-Type': 'application/json'})

else:
    verb = "GET"
    message = verb + " " + url

    # Sign the URL

    hmac_sha1 = hmac.new(sharedKey.encode(), message.encode(), digestmod=hashlib.sha1)
    digest = base64.b64encode(hmac_sha1.digest())

    # Fetch the data

    final_url = url + "&apiKey=" + apiKey + "&digest=" + urllib.parse.quote(digest.decode())
    req = urllib.request.Request(final_url)


# Perform the request

with urllib.request.urlopen(req) as response:
    responseJson = response.read()

# Pretty print the data

responseObject = json.loads(responseJson.decode("utf-8"))
print(json.dumps(responseObject, sort_keys=True, indent=2, separators=(',', ': ')))
